core/shared/src/main/scala/scalaz/zio/stream/Stream.scala

```diff
@@ -0,0 +1,733 @@
+        def loop(s: S, sched: schedule.State): IO[E1, S] =
+          self.foldLazy[E1, A1, S](s)(cont)(f).seq(schedule.update((), sched, clock)).flatMap {
+            case (s, decision) =>
+              if (decision.cont) IO.unit.delay(decision.delay) *> loop(s, decision.state)
+              else IO.now(s)
+          }
+
+        schedule.initial(clock).flatMap(loop(s, _))
+      }
+    }
+
+  /**
+   * Repeats elements of the stream using the provided schedule.
+   */
+  def repeatElems[B](schedule: Schedule[A, B], clock: Clock = Clock.Live): Stream[E, A] =
+    new Stream[E, A] {
+      override def foldLazy[E1 >: E, A1 >: A, S](s: S)(cont: S => Boolean)(f: (S, A1) => IO[E1, S]) = {
+        def loop(s: S, sched: schedule.State, a: A): IO[E1, S] =
+          schedule.update(a, sched, clock).flatMap { decision =>
+            if (decision.cont)
+              IO.unit.delay(decision.delay) *> f(s, a).flatMap(loop(_, decision.state, a))
+            else IO.now(s)
+          }
+
+        schedule.initial(clock).flatMap { sched =>
+          self.foldLazy[E1, A, S](s)(cont) { (s, a) =>
+            loop(s, sched, a)
+          }
+        }
+      }
+    }
+
+  /**
+   * Runs the sink on the stream to produce either the sink's result or an error.
+   */
+  def run[E1 >: E, A0, A1 >: A, B](sink: Sink[E1, A0, A1, B]): IO[E1, B] =
+    sink.initial
+      .flatMap(
+        state =>
+          self
+            .foldLazy[E1, A1, Sink.Step[sink.State, A0]](state)(Sink.Step.cont)(
+              (s, a) => sink.step(Sink.Step.state(s), a)
+            )
+            .flatMap(step => sink.extract(Sink.Step.state(step)))
+      )
+
+  /**
+   * Statefully maps over the elements of this stream to produce new elements.
+   */
+  def scan[S1, B](s1: S1)(f1: (S1, A) => (S1, B)): Stream[E, B] = new Stream[E, B] {
+    override def fold[E1 >: E, B1 >: B, S](s: S)(f: (S, B1) => IO[E1, Step[S]]): IO[E1, Step[S]] =
+      self
+        .fold[E1, A, (S, S1)](s -> s1) {
+          case ((s, s1), a) =>
+            val (s2, b) = f1(s1, a)
+
+            f(s, b).map(_.map(s => s -> s2))
+        }
+        .map(_.map(_._1))
+  }
+
+  /**
+   * Statefully and effectfully maps over the elements of this stream to produce
+   * new elements.
+   */
+  final def scanM[E1 >: E, S1, B](s1: S1)(f1: (S1, A) => IO[E1, (S1, B)]): Stream[E1, B] = new Stream[E1, B] {
+    override def fold[E2 >: E1, B1 >: B, S](s: S)(f: (S, B1) => IO[E2, Step[S]]): IO[E2, Step[S]] =
+      self
+        .fold[E2, A, (S, S1)](s -> s1) {
+          case ((s, s1), a) =>
+            f1(s1, a).flatMap {
+              case (s1, b) =>
+                f(s, b).map(_.map(s => s -> s1))
+            }
+        }
+        .map(_.map(_._1))
+  }
+
+  /**
+   * Takes the specified number of elements from this stream.
+   */
+  final def take(n: Int): Stream[E, A] =
+    self.zipWithIndex.takeWhile(_._2 == n - 1).map(_._1)
```

Comment:
Shouldn't it be `self.zipWithIndex.takeWhile(_._2 < n).map(_._1)` ?

Reply:
Yes, thank you